File: csharp/AssemblerWebJs/AssemblerEndpoint.cs
using Assembler.Api;
using Assembler.Common;
using Assembler.Config;
using Assembler.Engine;
using Assembler.Loader;
using Assembler.Performance;
using Assembler.Test;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace AssemblerWebJs
{
    #region Model and Serialization

    // Model for scenario response
    public class ScenarioDto
    {
        [System.Text.Json.Serialization.JsonPropertyName("appSite")]
        public string AppSite { get; set; } = string.Empty;
        [System.Text.Json.Serialization.JsonPropertyName("appFile")]
        public string AppFile { get; set; } = string.Empty;
        [System.Text.Json.Serialization.JsonPropertyName("appView")]
        public string AppView { get; set; } = string.Empty;
        [System.Text.Json.Serialization.JsonPropertyName("displayName")]
        public string DisplayName { get; set; } = string.Empty;
        [System.Text.Json.Serialization.JsonPropertyName("description")]
        public string Description { get; set; } = string.Empty;
    }

    // Model for merge request
    public class MergeRequest
    {
        [System.Text.Json.Serialization.JsonPropertyName("appSite")]
        public string? AppSite { get; set; }
        [System.Text.Json.Serialization.JsonPropertyName("appView")]
        public string? AppView { get; set; }
        [System.Text.Json.Serialization.JsonPropertyName("engineType")]
        public string? EngineType { get; set; }
    }
    
    [JsonSerializable(typeof(ScenarioDto[]))]
    [JsonSerializable(typeof(MergeRequest))]
    public partial class AssemblerJsonContext : JsonSerializerContext { }

    #endregion

    public static class AssemblerEndpoint
    {
        public static void MapAssemblerEndpoints(this WebApplication app)
        // POST endpoint for merging templates
        {
            var assemblerGroup = app.MapGroup("")
                .WithTags("Assembler");

            #region Root Endpoint

            assemblerGroup.MapGet("/", (HttpContext context) =>
            {
                // Use Index AppSite with engine toggle parameter
                string rootDirPath = Path.Combine(context.RequestServices.GetRequiredService<IHostEnvironment>().ContentRootPath, "wwwroot");

                // Get engine type from query parameter (default to Normal)
                var engineType = context.Request.Query["engine"].ToString();
                if (string.IsNullOrEmpty(engineType))
                {
                    engineType = "Normal";
                }

                // Validate EngineType against allowlist
                if (!SecurityValidator.ValidEngineTypes.Contains(engineType))
                    return Results.BadRequest("Invalid engine type. Use 'Normal' or 'PreProcess'");

                // Load templates for Index AppSite
                var normalTemplatesRaw = LoaderNormal.LoadGetTemplateFiles(rootDirPath, "Index");
                var preprocessTemplatesRaw = LoaderPreProcess.LoadProcessGetTemplateFiles(rootDirPath, "Index");

                // Merge using selected engine (no AppView context for Index)
                string mergedHtml = "";
                if (engineType.Equals("PreProcess", StringComparison.OrdinalIgnoreCase))
                {
                    var engine = new EnginePreProcess();
                    mergedHtml = engine.MergeTemplates("Index", "Index", null, preprocessTemplatesRaw.Templates);
                }
                else
                {
                    var engine = new EngineNormal();
                    mergedHtml = engine.MergeTemplates("Index", "Index", null, normalTemplatesRaw);
                }

                return Results.Content(mergedHtml, "text/html");
            })
            .WithName("GetRootUrl")
            .WithDisplayName("Get Method to Test Merging with Index AppSite")
            .WithDescription("Get Method to Test Merging with Index AppSite - use ?engine=Normal or ?engine=PreProcess")
            .WithTags("Root");

            #endregion

            #region Get Scenarios Endpoint

            assemblerGroup.MapGet("/api/scenarios", (HttpContext context) =>
            {
                try
                {
                    var scenarios = ConfigUtil.GetScenarios();
                    var scenarioDtos = scenarios.Select(s => new ScenarioDto
                    {
                        AppSite = s.AppSite,
                        AppFile = s.AppFile,
                        AppView = s.AppView,
                        DisplayName = s.DisplayName,
                        Description = s.Description
                    }).ToArray();

                    return Results.Json(scenarioDtos, AssemblerJsonContext.Default.ScenarioDtoArray);
                }
                catch (Exception ex)
                {
                    return Results.Problem($"Error loading scenarios: {ex.Message}");
                }
            })
            .WithName("GetScenarios")
            .WithDisplayName("Get Scenarios")
            .WithDescription("Returns all available scenarios from scenarios.csv")
            .WithTags("Scenarios");

            #endregion

            #region Merge Endpoint

            assemblerGroup.MapPost("/merge", async (HttpContext context) =>
            {
                var serverStart = DateTime.UtcNow;

                // Enable logging for merge operations
                var originalLogLevel = Logger.GetLogLevel();
                string projectDirectory = context.RequestServices.GetRequiredService<IHostEnvironment>().ContentRootPath;

                var templateAnalysisDir = Path.Combine(projectDirectory, "template_analysis");
                var logsDir = Path.Combine(templateAnalysisDir, "logs");
                Directory.CreateDirectory(logsDir);

                var contextLogFiles = new Dictionary<string, string>
                {
                    { "LoaderNormal", Path.Combine(logsDir, "csharp_loadernormal.log") },
                    { "LoaderPreProcess", Path.Combine(logsDir, "csharp_loaderpreprocess.log") },
                    { "EngineNormal", Path.Combine(logsDir, "csharp_enginenormal.log") },
                    { "EnginePreProcess", Path.Combine(logsDir, "csharp_enginepreprocess.log") }
                };

                Logger.Configure(Logger.LogLevel.DEBUG, null, false);
                Logger.ConfigureContextLogFiles(contextLogFiles);

                try
                {
                    using var reader = new StreamReader(context.Request.Body);
                    var body = await reader.ReadToEndAsync();
                    if (string.IsNullOrWhiteSpace(body))
                        return Results.BadRequest("Empty request body");

                    var input = System.Text.Json.JsonSerializer.Deserialize<MergeRequest>(body, AssemblerJsonContext.Default.MergeRequest);
                    if (input == null || string.IsNullOrWhiteSpace(input.AppSite) || string.IsNullOrWhiteSpace(input.EngineType))
                        return Results.BadRequest("Missing required fields: AppSite, EngineType");

                    string rootDirPath = Path.Combine(context.RequestServices.GetRequiredService<IHostEnvironment>().ContentRootPath, "wwwroot");

                    // Validate AppSite against allowlist loaded from appsites.csv
                    var validAppSites = SecurityValidator.GetValidAppSites();
                    if (!validAppSites.Contains(input.AppSite))
                        return Results.BadRequest("Invalid AppSite value");

                    // Validate EngineType against allowlist
                    if (!SecurityValidator.ValidEngineTypes.Contains(input.EngineType))
                        return Results.BadRequest("Invalid EngineType value");

                    // Validate path components for path traversal attacks
                    if (!SecurityValidator.IsValidPathComponent(input.AppSite))
                        return Results.BadRequest("Invalid characters in AppSite");

                    if (!string.IsNullOrEmpty(input.AppView) && !SecurityValidator.IsValidPathComponent(input.AppView))
                        return Results.BadRequest("Invalid characters in AppView");

                    // Get AppFile from scenarios
                    var scenarios = ConfigUtil.GetScenarios();
                    var appView = input.AppView ?? "";
                    var matchingScenario = scenarios.FirstOrDefault(s =>
                        s.AppSite.Equals(input.AppSite, StringComparison.OrdinalIgnoreCase) &&
                        s.AppView.Equals(appView, StringComparison.OrdinalIgnoreCase));

                    if (matchingScenario == null)
                        return Results.BadRequest($"No matching scenario found for AppSite='{input.AppSite}' and AppView='{appView}'");

                    var appFile = matchingScenario.AppFile;
                    var appViewPrefix = string.IsNullOrEmpty(appView) ? "" : appFile;

                    var normalTemplatesRaw = LoaderNormal.LoadGetTemplateFiles(rootDirPath, input.AppSite);
                    var preprocessTemplatesRaw = LoaderPreProcess.LoadProcessGetTemplateFiles(rootDirPath, input.AppSite);

                    var normalResult = normalTemplatesRaw.ToDictionary(
                        kvp => kvp.Key,
                        kvp => new TemplateData { Html = kvp.Value.html, Json = kvp.Value.json }
                    );

                    var preprocessResult = preprocessTemplatesRaw.Templates.ToDictionary(
                        kvp => kvp.Key,
                        kvp => new PreProcessTemplateMetadata
                        {
                            OriginalContent = kvp.Value.OriginalContent,
                            Placeholders = kvp.Value.Placeholders,
                            SlottedTemplates = kvp.Value.SlottedTemplates,
                            JsonData = kvp.Value.JsonData,
                            JsonPlaceholders = kvp.Value.JsonPlaceholders,
                            ReplacementMappings = kvp.Value.ReplacementMappings,
                            HasPlaceholders = kvp.Value.HasPlaceholders,
                            HasSlottedTemplates = kvp.Value.HasSlottedTemplates,
                            HasJsonData = kvp.Value.HasJsonData,
                            HasJsonPlaceholders = kvp.Value.HasJsonPlaceholders,
                            HasReplacementMappings = kvp.Value.HasReplacementMappings,
                            RequiresProcessing = kvp.Value.RequiresProcessing
                        }
                    );

                    var engineStart = DateTime.UtcNow;
                    string mergedHtml = "";
                    if (input.EngineType.Equals("PreProcess", System.StringComparison.OrdinalIgnoreCase))
                    {
                        var engine = new EnginePreProcess();
                        if (!string.IsNullOrEmpty(appViewPrefix))
                            engine.AppViewPrefix = appViewPrefix;
                        mergedHtml = engine.MergeTemplates(input.AppSite, appFile, appView, preprocessTemplatesRaw.Templates);
                    }
                    else
                    {
                        var engine = new EngineNormal();
                        if (!string.IsNullOrEmpty(appViewPrefix))
                            engine.AppViewPrefix = appViewPrefix;
                        mergedHtml = engine.MergeTemplates(input.AppSite, appFile, appView, normalTemplatesRaw);
                    }
                    var engineTimeMs = (DateTime.UtcNow - engineStart).TotalMilliseconds;
                    var serverEnd = DateTime.UtcNow;
                    var serverTimeMs = (serverEnd - serverStart).TotalMilliseconds;

                    // Save HTML output only if save query parameter is present
                    var saveParam = context.Request.Query["save"].ToString();
                    if (!string.IsNullOrEmpty(saveParam) && saveParam.Equals("true", StringComparison.OrdinalIgnoreCase))
                    {
                        var contentRoot = context.RequestServices.GetRequiredService<IHostEnvironment>().ContentRootPath;
                        var outputDir = Path.Combine(contentRoot, "template_analysis", "output");
                        Directory.CreateDirectory(outputDir);

                        var appViewSuffix = string.IsNullOrEmpty(input.AppView) ? "" : $"_{input.AppView}";
                        var engineSuffix = input.EngineType.ToLower();
                        var outputFile = Path.Combine(outputDir, $"{input.AppSite}{appViewSuffix}_{engineSuffix}.html");
                        File.WriteAllText(outputFile, mergedHtml);

                        // Also generate the structure dump file, filtered by AppSite
                        var allScenarios = ConfigUtil.GetScenarios();
                        var filteredScenarios = ConfigUtil.FilterByAppSite(allScenarios, input.AppSite);
                        TestingUtils.DumpPreprocessedTemplateStructures(rootDirPath, contentRoot, filteredScenarios, true);
                    }

                    var responseObj = new ApiResponse
                    {
                        Templates = normalResult,
                        PreProcessTemplates = preprocessResult,
                        AppSite = input.AppSite ?? string.Empty,
                        AppFile = appFile,
                        AppView = appView,
                        ServerTimeMs = serverTimeMs,
                        Html = mergedHtml,
                        EngineTimeMs = engineTimeMs
                    };
                    var responseJson = responseObj.SerializeToJson();

                    return Results.Content(responseJson, "application/json");
                }
                finally
                {
                    // Restore original log level
                    Logger.SetLogLevel(originalLogLevel);
                }
            })
            .Accepts<MergeRequest>("application/json")
            .WithName("PostMergeTemplate")
            .WithDisplayName("Post Method to Merge Template for AppSite, AppView, EngineType")
            .WithDescription("Post Method to Merge Template for AppSite, AppView, EngineType. AppFile is retrieved from scenarios.")
            .WithTags("Merge");

            #endregion

            #region Get Templates Endpoint

            assemblerGroup.MapPost("/api/templates", async (HttpContext context) =>
            {
                var serverStart = DateTime.UtcNow;

                // Enable logging for merge operations
                var originalLogLevel = Logger.GetLogLevel();
                string projectDirectory = context.RequestServices.GetRequiredService<IHostEnvironment>().ContentRootPath;

                var templateAnalysisDir = Path.Combine(projectDirectory, "template_analysis");
                var logsDir = Path.Combine(templateAnalysisDir, "logs");
                Directory.CreateDirectory(logsDir);

                var contextLogFiles = new Dictionary<string, string>
                    {
                        { "LoaderNormal", Path.Combine(logsDir, "csharp_loadernormal.log") },
                        { "LoaderPreProcess", Path.Combine(logsDir, "csharp_loaderpreprocess.log") }
                    };

                Logger.Configure(Logger.LogLevel.DEBUG, null, false);
                Logger.ConfigureContextLogFiles(contextLogFiles);

                try
                {
                    string rootDirPath = Path.Combine(context.RequestServices.GetRequiredService<IHostEnvironment>().ContentRootPath, "wwwroot");

                    // Read the request body
                    using var reader = new StreamReader(context.Request.Body);
                    var requestBody = await reader.ReadToEndAsync();

                    // Parse JSON manually for NativeAOT compatibility
                    var appSite = "";
                    try
                    {
                        var jsonDoc = System.Text.Json.JsonDocument.Parse(requestBody);
                        if (jsonDoc.RootElement.TryGetProperty("appsite", out var appSiteElement))
                        {
                            appSite = appSiteElement.GetString() ?? "";
                        }
                    }
                    catch
                    {
                        return Results.Text("Invalid JSON format", statusCode: 400);
                    }

                    if (string.IsNullOrEmpty(appSite))
                        return Results.Text("Missing appsite parameter", statusCode: 400);

                    // Validate AppSite against allowlist loaded from appsites.csv
                    var validAppSites = SecurityValidator.GetValidAppSites();
                    if (!validAppSites.Contains(appSite))
                        return Results.Text("Invalid AppSite value", statusCode: 400);

                    // Validate path components for path traversal attacks
                    if (!SecurityValidator.IsValidPathComponent(appSite))
                        return Results.Text("Invalid characters in AppSite", statusCode: 400);
                  
                    // Load Normal templates
                    var normalTemplates = LoaderNormal.LoadGetTemplateFiles(rootDirPath, appSite);

                    // Load PreProcess templates
                    var preprocessTemplates = LoaderPreProcess.LoadProcessGetTemplateFiles(rootDirPath, appSite);

                    // Convert Normal templates to TemplateData objects for proper JSON serialization
                    var normalResult = normalTemplates.ToDictionary(
                        kvp => kvp.Key,
                        kvp => new TemplateData { Html = kvp.Value.html, Json = kvp.Value.json }
                    );

                    // Convert PreProcess templates to metadata-only objects
                    var preprocessResult = preprocessTemplates.Templates.ToDictionary(
                        kvp => kvp.Key,
                        kvp => new PreProcessTemplateMetadata
                        {
                            OriginalContent = kvp.Value.OriginalContent,
                            Placeholders = kvp.Value.Placeholders,
                            SlottedTemplates = kvp.Value.SlottedTemplates,
                            JsonData = kvp.Value.JsonData,
                            JsonPlaceholders = kvp.Value.JsonPlaceholders,
                            ReplacementMappings = kvp.Value.ReplacementMappings,
                            HasPlaceholders = kvp.Value.HasPlaceholders,
                            HasSlottedTemplates = kvp.Value.HasSlottedTemplates,
                            HasJsonData = kvp.Value.HasJsonData,
                            HasJsonPlaceholders = kvp.Value.HasJsonPlaceholders,
                            HasReplacementMappings = kvp.Value.HasReplacementMappings,
                            RequiresProcessing = kvp.Value.RequiresProcessing
                        }
                    );

                    var serverEnd = DateTime.UtcNow;
                    var serverTimeMs = (serverEnd - serverStart).TotalMilliseconds;

                    // Use named response class for NativeAOT compatibility
                    var response = new ApiResponse
                    {
                        Templates = normalResult,
                        PreProcessTemplates = preprocessResult,
                        AppSite = appSite,
                        AppFile = null,
                        AppView = null,
                        ServerTimeMs = serverTimeMs
                    };

                    var jsonResult = response.SerializeToJson();

                    // Check if save query parameter is present
                    var saveParam = context.Request.Query["save"].ToString();
                    if (!string.IsNullOrEmpty(saveParam) && saveParam.Equals("true", StringComparison.OrdinalIgnoreCase))
                    {
                        var templatesDir = Path.Combine(projectDirectory, "template_analysis", "templates");
                        Directory.CreateDirectory(templatesDir);

                        var saveFile = Path.Combine(templatesDir, $"csharp_{appSite}_templates.json");
                        await File.WriteAllTextAsync(saveFile, jsonResult);

                        // Save structure dump using TestingUtils logic
                        // Build a scenario list for the requested appSite
                        var scenarios = new List<Assembler.Config.Scenario> {
                            new Assembler.Config.Scenario(appSite, "", "")
                        };
                        Assembler.Test.TestingUtils.DumpPreprocessedTemplateStructures(rootDirPath, projectDirectory, scenarios, true);
                    }

                    return Results.Content(jsonResult, "application/json");
                }
                catch (Exception ex)
                {
                    return Results.Text($"Error: {ex.Message}", statusCode: 500);
                }
                finally
                {
                    // Restore original log level
                    Logger.SetLogLevel(originalLogLevel);
                }
            });

            #endregion
        }
    }
}

File: php/AssemblerWeb/AssemblerEndpoint.php
<?php

require_once __DIR__ . '/SecurityValidator.php';

use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as ServerRequest;
use Assembler\Engine\EngineNormal;
use Assembler\Engine\EnginePreProcess;
use Assembler\Loader\LoaderNormal;
use Assembler\Loader\LoaderPreProcess;
use Assembler\Api\ApiResponse;
use Assembler\Api\TemplateData;
use Assembler\Api\PreProcessTemplateMetadata;
use Assembler\Common\Logger;
use Assembler\Test\TestingUtils;
use Assembler\Performance\PerformanceUtils;
use Assembler\Config\ConfigUtil;

class AssemblerEndpoint
{
    public static function indexEndpoint(ServerRequest $request, Response $response): Response
    {
        try {
            $rootDirPath = __DIR__ . DIRECTORY_SEPARATOR . 'wwwroot';

            $queryParams = $request->getQueryParams();
            $engineType = $queryParams['engine'] ?? 'Normal';

            if (!SecurityValidator::isValidEngineType($engineType)) {
                $response->getBody()->write('Invalid engine type. Use \'Normal\' or \'PreProcess\'');
                return $response->withStatus(400);
            }

            LoaderNormal::clearCache();
            LoaderPreProcess::clearCache();

            $normalTemplatesRaw = LoaderNormal::loadGetTemplateFiles($rootDirPath, 'Index');
            $preprocessTemplatesRaw = LoaderPreProcess::loadProcessGetTemplateFiles($rootDirPath, 'Index');

            $mergedHtml = '';
            if (strcasecmp($engineType, 'PreProcess') === 0) {
                $engine = new EnginePreProcess();
                $mergedHtml = $engine->mergeTemplates('Index', 'Index', null, $preprocessTemplatesRaw->templates);
            } else {
                $engine = new EngineNormal();
                $mergedHtml = $engine->mergeTemplates('Index', 'Index', null, $normalTemplatesRaw);
            }

            $response->getBody()->write($mergedHtml);
            return $response->withHeader('Content-Type', 'text/html');
        } catch (Exception $error) {
            error_log('Error in root endpoint: ' . $error->getMessage());
            $response->getBody()->write('Internal server error');
            return $response->withStatus(500);
        }
    }

	public static function scenariosEndpoint(ServerRequest $request, Response $response): Response
    {
        try {
            $scenarios = ConfigUtil::getScenarios();
            $scenarioDtos = [];

            foreach ($scenarios as $s) {
                $scenarioDtos[] = [
                    'appSite' => $s->appSite,
                    'appFile' => $s->appFile,
                    'appView' => $s->appView,
                    'displayName' => $s->displayName,
                    'description' => $s->description
                ];
            }

            $response->getBody()->write(json_encode($scenarioDtos));
            return $response->withHeader('Content-Type', 'application/json');
        } catch (Exception $error) {
            error_log('Error getting scenarios: ' . $error->getMessage());
            $response->getBody()->write(json_encode(['error' => 'Error loading scenarios: ' . $error->getMessage()]));
            return $response->withStatus(500)->withHeader('Content-Type', 'application/json');
        }
    }

    public static function mergeEndpoint(ServerRequest $request, Response $response): Response
    {
        // Enable logging for merge operations
        $originalLogLevel = Logger::getLogLevel();

        $projectRootPath = dirname(__DIR__, 1);
        $templateAnalysisDir = $projectRootPath . DIRECTORY_SEPARATOR . 'template_analysis';
        $logsDir = $templateAnalysisDir . DIRECTORY_SEPARATOR . 'logs';
        if (!is_dir($logsDir)) {
            mkdir($logsDir, 0755, true);
        }

        $contextLogFiles = [
            'LoaderNormal' => $logsDir . DIRECTORY_SEPARATOR . 'php_loadernormal.log',
            'LoaderPreProcess' => $logsDir . DIRECTORY_SEPARATOR . 'php_loaderpreprocess.log',
            'EngineNormal' => $logsDir . DIRECTORY_SEPARATOR . 'php_enginenormal.log',
            'EnginePreProcess' => $logsDir . DIRECTORY_SEPARATOR . 'php_enginepreprocess.log'
        ];

        Logger::configure(Logger::DEBUG, null, false);
        Logger::configureContextLogFiles($contextLogFiles);

        $serverStart = microtime(true) * 1000;

        try {
            $body = $request->getBody()->getContents();
            $data = json_decode($body, true);

            if (json_last_error() !== JSON_ERROR_NONE) {
                Logger::setLogLevel($originalLogLevel);
                $response->getBody()->write(json_encode(['error' => 'Invalid JSON']));
                return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
            }

            $appSite = $data['appSite'] ?? null;
            $appView = $data['appView'] ?? null;
            $engineType = $data['engineType'] ?? null;

            if (empty($appSite) || empty($engineType)) {
                Logger::setLogLevel($originalLogLevel);
                $response->getBody()->write(json_encode(['error' => 'Missing required fields: appSite, engineType']));
                return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
            }

            // Get AppFile from scenarios
            $scenarios = ConfigUtil::getScenarios();
            $appViewValue = $appView ?? '';
            $matchingScenario = null;

            foreach ($scenarios as $s) {
                if (strcasecmp($s->appSite, $appSite) === 0 && strcasecmp($s->appView, $appViewValue) === 0) {
                    $matchingScenario = $s;
                    break;
                }
            }

            if ($matchingScenario === null) {
                Logger::setLogLevel($originalLogLevel);
                $response->getBody()->write(json_encode(['error' => "No matching scenario found for AppSite='$appSite' and AppView='$appViewValue'"]));
                return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
            }

            $appFile = $matchingScenario->appFile;

            // Calculate appViewPrefix from appFile when appView is not empty
            $appViewPrefix = (!empty($appView)) ? $appFile : '';

            $logMsg = sprintf(
                "/merge endpoint called with: app_site=%s, app_file=%s, engine_type=%s, app_view=%s, app_view_prefix=%s",
                $appSite ?? 'null',
                $appFile ?? 'null',
                $engineType ?? 'null',
                $appView ?? 'null',
                $appViewPrefix ?? 'null'
            );
            error_log($logMsg);
            Logger::info($logMsg, 'MergeEndpoint');

            $rootDirPath = __DIR__ . DIRECTORY_SEPARATOR . 'wwwroot';

            if (!SecurityValidator::isValidEngineType($engineType)) {
                Logger::setLogLevel($originalLogLevel);
                $response->getBody()->write(json_encode(['error' => 'Invalid EngineType value']));
                return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
            }

            $validAppSites = [];
            try {
                $validAppSites = SecurityValidator::getValidAppSites($rootDirPath);
            } catch (Exception $error) {
                Logger::setLogLevel($originalLogLevel);
                $response->getBody()->write(json_encode(['error' => 'Failed to load AppSites: ' . $error->getMessage()]));
                return $response->withStatus(500)->withHeader('Content-Type', 'application/json');
            }

            if (!SecurityValidator::isValidAppSite($appSite, $validAppSites)) {
                Logger::setLogLevel($originalLogLevel);
                $response->getBody()->write(json_encode(['error' => 'Invalid AppSite value']));
                return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
            }

            if (!SecurityValidator::isValidPathComponent($appSite)) {
                Logger::setLogLevel($originalLogLevel);
                $response->getBody()->write(json_encode(['error' => 'Invalid characters in AppSite']));
                return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
            }

            if (!SecurityValidator::isValidPathComponent($appFile)) {
                Logger::setLogLevel($originalLogLevel);
                $response->getBody()->write(json_encode(['error' => 'Invalid characters in AppFile']));
                return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
            }

            if ($appView !== null && $appView !== '' && !SecurityValidator::isValidPathComponent($appView)) {
                Logger::setLogLevel($originalLogLevel);
                $response->getBody()->write(json_encode(['error' => 'Invalid characters in AppView']));
                return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
            }

            LoaderNormal::clearCache();
            LoaderPreProcess::clearCache();

            $templatesMap = LoaderNormal::loadGetTemplateFiles($rootDirPath, $appSite);
            $preTemplates = LoaderPreProcess::loadProcessGetTemplateFiles($rootDirPath, $appSite);

            $apiResponse = new ApiResponse();
            $apiResponse->appSite = $appSite;
            $apiResponse->appFile = $appFile;
            $apiResponse->appView = $appView;

            foreach ($templatesMap as $key => $value) {
                $templateData = new TemplateData();
                $templateData->html = $value->html ?? '';
                $templateData->json = $value->json ?? null;
                $apiResponse->templates[$key] = $templateData;
            }

            foreach ($preTemplates->templates as $key => $value) {
                $metadata = new PreProcessTemplateMetadata();
                $metadata->originalContent = $value->originalContent;
                $metadata->placeholders = $value->placeholders;
                $metadata->slottedTemplates = $value->slottedTemplates;
                $metadata->jsonData = $value->jsonData;
                $metadata->jsonPlaceholders = $value->jsonPlaceholders;
                $metadata->replacementMappings = $value->replacementMappings;
                $metadata->hasPlaceholders = $value->hasPlaceholders();
                $metadata->hasSlottedTemplates = $value->hasSlottedTemplates();
                $metadata->hasJsonData = $value->hasJsonData();
                $metadata->hasJsonPlaceholders = $value->hasJsonPlaceholders();
                $metadata->hasReplacementMappings = $value->hasReplacementMappings();
                $metadata->requiresProcessing = $value->requiresProcessing();
                $apiResponse->preProcessTemplates[$key] = $metadata;
            }

            $engineStart = microtime(true) * 1000;

            $mergedHtml = '';
            if (strcasecmp($engineType, 'PreProcess') === 0) {
                $engine = new EnginePreProcess();
                if (!empty($appViewPrefix)) {
                    $engine->setAppViewPrefix($appViewPrefix);
                }
                $mergedHtml = $engine->mergeTemplates($appSite, $appFile, $appView, $preTemplates->templates);
                $apiResponse->templates = [];
            } else {
                $engine = new EngineNormal();
                if (!empty($appViewPrefix)) {
                    $engine->setAppViewPrefix($appViewPrefix);
                }
                $mergedHtml = $engine->mergeTemplates($appSite, $appFile, $appView, $templatesMap);
                $apiResponse->preProcessTemplates = [];
            }

            $apiResponse->engineTimeMs = (microtime(true) * 1000) - $engineStart;
            $apiResponse->serverTimeMs = (microtime(true) * 1000) - $serverStart;
            $apiResponse->html = $mergedHtml;

            // Save HTML output only if save query parameter is present
            $queryParams = $request->getQueryParams();
            $saveParam = $queryParams['save'] ?? '';
            if (strcasecmp($saveParam, 'true') === 0) {
                $outputDir = $projectRootPath . DIRECTORY_SEPARATOR . 'template_analysis' . DIRECTORY_SEPARATOR . 'output';
                if (!is_dir($outputDir)) {
                    mkdir($outputDir, 0755, true);
                }

                $appViewSuffix = (!empty($appView)) ? "_{$appView}" : '';
                $engineSuffix = strtolower($engineType);
                $outputFile = $outputDir . DIRECTORY_SEPARATOR . "{$appSite}{$appViewSuffix}_{$engineSuffix}.html";
                file_put_contents($outputFile, $mergedHtml);
            }

            // Restore original log level
            Logger::setLogLevel($originalLogLevel);

            $response->getBody()->write($apiResponse->serializeToJson());
            return $response->withHeader('Content-Type', 'application/json');
        } catch (Exception $error) {
            Logger::setLogLevel($originalLogLevel);
            error_log('Error in merge endpoint: ' . $error->getMessage());
            $response->getBody()->write(json_encode(['error' => 'Internal server error']));
            return $response->withStatus(500)->withHeader('Content-Type', 'application/json');
        }
    }

	public static function getTemplatesEndpoint(ServerRequest $request, Response $response): Response
    {
        try {
            $rootDirPath = __DIR__ . DIRECTORY_SEPARATOR . 'wwwroot';
            $projectDirectory = dirname(__DIR__, 1);

            // Read request body
            $requestBody = (string)$request->getBody();
            error_log('[/api/templates] Raw request body: ' . $requestBody);
            
            $body = json_decode($requestBody, true);
            if (json_last_error() !== JSON_ERROR_NONE) {
                error_log('[/api/templates] JSON decode error: ' . json_last_error_msg());
                $response->getBody()->write('Invalid JSON format');
                return $response->withStatus(400);
            }
            
            error_log('[/api/templates] Decoded body: ' . print_r($body, true));
            $appSite = $body['appsite'] ?? '';
            error_log('[/api/templates] Extracted appSite: ' . $appSite);

            if (empty($appSite)) {
                error_log('[/api/templates] Missing appsite parameter');
                $response->getBody()->write('Missing appsite parameter');
                return $response->withStatus(400);
            }

            // Validate AppSite against allowlist loaded from appsites.csv
            $validAppSites = SecurityValidator::getValidAppSites($rootDirPath);
            error_log('[/api/templates] Valid AppSites: ' . print_r($validAppSites, true));
            // Case-insensitive comparison to match C# behavior (using array_map to lowercase for comparison)
            $validAppSitesLower = array_map('strtolower', $validAppSites);
            if (!in_array(strtolower($appSite), $validAppSitesLower, true)) {
                error_log('[/api/templates] AppSite validation failed for: ' . $appSite);
                $response->getBody()->write('Invalid AppSite value');
                return $response->withStatus(400);
            }
            error_log('[/api/templates] AppSite validation passed');

            // Validate path components for path traversal attacks
            if (!SecurityValidator::isValidPathComponent($appSite)) {
                $response->getBody()->write('Invalid characters in AppSite');
                return $response->withStatus(400);
            }

            $serverStart = microtime(true) * 1000;

            // Load Normal templates
            $normalTemplates = LoaderNormal::loadGetTemplateFiles($rootDirPath, $appSite);

            // Load PreProcess templates
            $preprocessTemplates = LoaderPreProcess::loadProcessGetTemplateFiles($rootDirPath, $appSite);

            // Convert Normal templates to TemplateData objects for proper JSON serialization
            $normalResult = [];
            foreach ($normalTemplates as $key => $value) {
                $templateData = new TemplateData();
                $templateData->html = $value->html ?? '';
                $templateData->json = $value->json ?? null;
                $normalResult[$key] = $templateData;
            }

            // Convert PreProcess templates to metadata-only objects
            $preprocessResult = [];
            foreach ($preprocessTemplates->templates as $key => $value) {
                $metadata = new PreProcessTemplateMetadata();
                $metadata->originalContent = $value->originalContent;
                $metadata->placeholders = $value->placeholders;
                $metadata->slottedTemplates = $value->slottedTemplates;
                $metadata->jsonData = $value->jsonData;
                $metadata->jsonPlaceholders = $value->jsonPlaceholders;
                $metadata->replacementMappings = $value->replacementMappings;
                $metadata->hasPlaceholders = $value->hasPlaceholders();
                $metadata->hasSlottedTemplates = $value->hasSlottedTemplates();
                $metadata->hasJsonData = $value->hasJsonData();
                $metadata->hasJsonPlaceholders = $value->hasJsonPlaceholders();
                $metadata->hasReplacementMappings = $value->hasReplacementMappings();
                $metadata->requiresProcessing = $value->requiresProcessing();
                $preprocessResult[$key] = $metadata;
            }

            $serverEnd = microtime(true) * 1000;
            $serverTimeMs = $serverEnd - $serverStart;

            // Use proper ApiResponse structure
            $apiResponse = new ApiResponse();
            $apiResponse->templates = $normalResult;
            $apiResponse->preProcessTemplates = $preprocessResult;
            $apiResponse->appSite = $appSite;
            $apiResponse->appFile = null;
            $apiResponse->appView = null;
            $apiResponse->serverTimeMs = $serverTimeMs;

            $jsonResult = $apiResponse->serializeToJson(false);

            // Check if save query parameter is present
            $queryParams = $request->getQueryParams();
            $saveParam = $queryParams['save'] ?? '';
            if (strcasecmp($saveParam, 'true') === 0) {
                $templatesDir = $projectDirectory . DIRECTORY_SEPARATOR . 'template_analysis' . DIRECTORY_SEPARATOR . 'templates';
                if (!is_dir($templatesDir)) {
                    mkdir($templatesDir, 0755, true);
                }

                $saveFile = $templatesDir . DIRECTORY_SEPARATOR . "php_{$appSite}_templates.json";
                file_put_contents($saveFile, $jsonResult);

                // Save structure dump using TestingUtils logic
                // Build a scenario list for the requested appSite
                $scenarios = [(object)['appSite' => $appSite, 'appFile' => '', 'appView' => '']];
                TestingUtils::dumpPreprocessedTemplateStructures($rootDirPath, $projectDirectory, $scenarios, true);
            }

            $response->getBody()->write($jsonResult);
            return $response->withHeader('Content-Type', 'application/json');
        } catch (Exception $error) {
            error_log('Error in /api/templates: ' . $error->getMessage());
            $response->getBody()->write('Error: ' . $error->getMessage());
            return $response->withStatus(500);
        }
    }

}
